server/src/main/java/org/opensearch/indices/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,264 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.indices;
+
+import org.opensearch.common.metrics.CounterMetric;
+import org.roaringbitmap.RoaringBitmap;
+
+import java.util.HashSet;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * This class implements KeyLookupStore<Integer> using a roaring bitmap with a modulo applied to values.
+ * The modulo increases the density of values, which makes RBMs more memory-efficient. The recommended modulo is ~2^28.
+ * It also maintains a hash set of values which have had collisions. Values which haven't had collisions can be
+ * safely removed from the store. The fraction of collided values should be low,
+ * about 0.5% for a store with 10^7 values and a modulo of 2^28.
+ * The store estimates its memory footprint and will stop adding more values once it reaches its memory cap.
+ */
+public class RBMIntKeyLookupStore implements KeyLookupStore<Integer> {
+    protected final int modulo;
+    protected class KeyStoreStats {
+        protected int size;
+        protected long memSizeCapInBytes;
+        protected CounterMetric numAddAttempts;
+        protected CounterMetric numCollisions;
+        protected boolean guaranteesNoFalseNegatives;
+        protected int maxNumEntries;
+        protected boolean atCapacity;
+        protected CounterMetric numRemovalAttempts;
+        protected CounterMetric numSuccessfulRemovals;
+        protected KeyStoreStats(long memSizeCapInBytes, int maxNumEntries) {
+            this.size = 0;
+            this.numAddAttempts = new CounterMetric();
+            this.numCollisions = new CounterMetric();
+            this.memSizeCapInBytes = memSizeCapInBytes;
+            this.maxNumEntries = maxNumEntries;
+            this.atCapacity = false;
+            this.numRemovalAttempts = new CounterMetric();
+            this.numSuccessfulRemovals = new CounterMetric();
+        }
+    }
+
+    protected KeyStoreStats stats;
+    protected RoaringBitmap rbm;
+    private HashSet<Integer> collidedInts;
+    protected RBMSizeEstimator sizeEstimator;
+    protected final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    protected final Lock readLock = lock.readLock();
+    protected final Lock writeLock = lock.writeLock();
+
+    RBMIntKeyLookupStore(int modulo, long memSizeCapInBytes) {
+        this.modulo = modulo;
+        sizeEstimator = new RBMSizeEstimator(modulo);
```

Comment:
Again if we confine modulo to specific set of values, we can calculate this and store it as part of some enum map.
As giving capability to pass memSizeCap seems wrong and error prone.

Reply:
Sounds good for the modulo. Why don't we want to pass the cap to the keystore?